adding ccloud::create_ccloud_stack_sub

gianlucanatali opened the pull request:
extracting logic from  ccloud::create_ccloud_stack into reusable function for my demo (ccloud::create_ccloud_stack_sub)

I have moved the definition of the variables up, so I can change them for my demo.
The name of the function is not great, happy to change that.
In general consider this a newbie contribution, so if anything sounds off let me know!!
Thank you @rspurgeon

gianlucanatali committed:
adding ccloud::create_ccloud_stack_sub

extracting logic from  ccloud::create_ccloud_stack into reusable function for my demo (ccloud::create_ccloud_stack_sub)

gianlucanatali committed:
adding ccloud::create_ccloud_stack_sub

extracting logic from  ccloud::create_ccloud_stack into reusable function for my demo (ccloud::create_ccloud_stack_sub)

rspurgeon commented:
@gianlucanatali I believe I understand the goal.  I think what I would prefer we do is make the variables in the existing function overridable instead of creating a new function entirely.

For example, if your goal is to be able to set KSQL_NAME from a call to `ccloud::create_ccloud_stack`, then we could assign `KSQL_NAME` with an optional default like so:

`KSQL_NAME=${KSQL_NAME:-"demo-ksql-$SERVICE_ACCOUNT_ID"}`

The same is true with `CLIENT_CONFIG`, if you desire to set that from your calling function, we can just setup a default like above.

I'm curious why you want to remove the `mkdir -p stack-configs` from the function that you're calling, are you storing the configs elsewhere?  If so, we could do a similar thing and make the path to the configs configurable with a default.

Thanks for your PR and let me know what you think about these proposed changes.  If you'd like some help w/ the bash code, please let me know.

gianlucanatali commented:
Hi @rspurgeon , thanks for quick feedback. Actually I am only using part of this function (the part I isolated in that function). I enter this function having already the environment and service account created. In my case I have 1 env, 1 service account, and 3 stacks inside. You can get a idea by looking at this code where I use the function:
https://github.com/gianlucanatali/confluent-multicloud-demo/blob/b8b8bc597adfbe35361f531c666cfb62f79013ea/create_demo_environment.sh#L218

This is thw starting point, here I actually use the function:
https://github.com/gianlucanatali/confluent-multicloud-demo/blob/b8b8bc597adfbe35361f531c666cfb62f79013ea/create_demo_environment.sh#L207

Thanks!

ybyzek commented on utils/ccloud_library.sh:
```diff
@@ -721,6 +721,16 @@ function ccloud::create_ccloud_stack() {
+  ccloud::create_ccloud_stack_sub
+}
+
+function ccloud::create_ccloud_stack_sub() {
```
Option 1: use the sub function but rename it to something like `create_ccloud_substack_per_env` ? (This may be the simplest option...just requires picking a good name)

Option 2: remove the entire idea of a sub function, and instead refactor `create_ccloud_stack` to use options, thus allowing user to pass in pre-existing service account id and environment to use.  (This is more complicated upfront but might give more freedom in the future)

cc: @rspurgeon WDYT?

rspurgeon replied:
As @gianlucanatali mentioned below, we discussed out of band making the function default variables instead of breaking the function up.  He's going to work on a new verison and push a change for that.

gianlucanatali committed:
making function ccloud::create_ccloud_stack()  more flexible

I have removed the _sub function and added the logic directly inside the original function, so that can reuse existing service account and environment.
The function now also accept name definition from the outside, defaulting to the demo names

gianlucanatali committed:
making function ccloud::create_ccloud_stack()  more flexible

I have removed the _sub function and added the logic directly inside the original function, so that can reuse existing service account and environment.
The function now also accept name definition from the outside, defaulting to the demo names

gianlucanatali commented:
Thanks @ybyzek and @rspurgeon fo help and guidance, I have changed the function using Option 2 as also discussed with Rick via slack, this seems a better approach.
I did not test the existing demo assets (other than mine) with this new version though.
Let me know what you think!

rspurgeon commented on utils/ccloud_library.sh:
```diff
@@ -708,29 +708,26 @@ function ccloud::set_kafka_cluster_use() {
 
-  SERVICE_NAME="demo-app-$RANDOM_NUM"
-  SERVICE_ACCOUNT_ID=$(ccloud::create_service_account $SERVICE_NAME)
   echo "Creating Confluent Cloud stack for new service account id $SERVICE_ACCOUNT_ID of name $SERVICE_NAME"
```
```suggestion
  echo "Creating Confluent Cloud stack for service account $SERVICE_NAME, ID: $SERVICE_ACCOUNT_ID.
```

gianlucanatali replied:
✅

rspurgeon commented on utils/ccloud_library.sh:
```diff
@@ -708,29 +708,26 @@ function ccloud::set_kafka_cluster_use() {
-  CLUSTER_NAME=demo-kafka-cluster-$SERVICE_ACCOUNT_ID
+  if [[ -z "$ENVIRONMENT" ]]; then
+    # Environment is not received so it will be created
+    echo "You need to specify a demo Type!"
```
I'm not sure what this print line is specifying?  It doesn't seem necessary

gianlucanatali replied:
yes, good catch, was a copy paste refuse. I will remove it

rspurgeon commented on utils/ccloud_library.sh:
```diff
@@ -708,29 +708,26 @@ function ccloud::set_kafka_cluster_use() {
+    local RANDOM_NUM=$((1 + RANDOM % 1000000))
+    SERVICE_NAME=${SERVICE_NAME:-"demo-app-$RANDOM_NUM"}
+    SERVICE_ACCOUNT_ID=$(ccloud::create_service_account $SERVICE_NAME)
+    
```
nit: remove empty line

gianlucanatali replied:
✅

ybyzek committed:
DEVX-1838: Ensure pageviews topic and subject exist before running KSQL (#650)

rspurgeon commented on utils/ccloud_library.sh:
```diff
@@ -708,29 +708,26 @@ function ccloud::set_kafka_cluster_use() {
+    SERVICE_NAME=${SERVICE_NAME:-"demo-app-$RANDOM_NUM"}
+    SERVICE_ACCOUNT_ID=$(ccloud::create_service_account $SERVICE_NAME)
+    
+  fi
```
After this line, I believe that we should check for empty SERVICE_NAME and exit with a message if it's not provided or wasn't calculated in the case of the SERVICE_ACCOUNT_ID being provided.

gianlucanatali replied:
Added the logic in, thx!

rspurgeon committed:
Removes a missing function from ccloud create stack (#651)

When the utils/helper.sh was removed as a source for this file
this function no longer became available.

rspurgeon committed:
Quick fix for quoted params (#652)

* Quotes the cluster name varible to ccloud kafka cluster create

If the input to this parameter isnt quoted, then it can fail if the parameter
has spaces.

* Quotes the ksql name variable as this name is allowed spaces.

rspurgeon commented:
@gianlucanatali Please rebase off 5.5.0-post to pickup some important changes related to these functions

gianlucanatali committed:
Merge branch 'gn-cloud-stack-sub-2' of github.com:gianlucanatali/examples into gn-cloud-stack-sub-2

gianlucanatali commented:
Hi @rspurgeon thanks a lot for your comments, I'll go ahead and change the function accordingly. I am not sure what I did wrong... I rebased off 5.5.0-post , and I now see on the changed files of this PR also files I don't think I changed ... Can you have a look?
I can also create a new PR and just bring with me the code in another branch but I think I need to figure out how to do this properly at some point :) Thanks!

gianlucanatali committed:
Updating ccloud_library.sh function ccloud::create_ccloud_stack()

gianlucanatali commented:
I pushed the changes to my branch, as I mentioned my intent is only to change the function function ccloud::create_ccloud_stack()  in ccloud_library.sh . But I do see other files changed in this PR, probably I screwed smg up in the rebase process.. guidance welcome :)
Thanks!

gianlucanatali commented:
Closing this PR , moving same changes to new one: https://github.com/confluentinc/examples/pull/656
